### ISSUE TITLE
Focus the current image in the library

### ISSUE THREAD

**Issue by maximbaz** [karlch/vimiv-qt#127]
Hi there, 

I'm looking into synchronizing what I see in the "image mode" with my cursor position in the "library mode".

When I'm in the "library mode", the `--open-selected` flag is awesome, I press `n` / `p` and the "image mode" always shows the image that is currently under cursor.

But I can't figure out how to achieve the opposite, let's say I open `vimiv img.jpg` and it opens in "image mode", and then I want to show the library sidebar and set my focus to the "img.jpg". If I simply press `tl`, the cursor in the library will be in the top. Am I missing an existing flag or a command to do so? If not, what do you think about implementing some form of `--open-selected` for `tl` as well?

**Comment by karlch** [comment]
This is a bit less straightforward as the image knows nothing about the library, while the library knows how to open images. So the image cannot tell the library, to focus the current path.

What would be possible is that the library always follows the images by connecting to `new_image_opened`. The behaviour would then be equivalent to thumbnail mode, so maybe this is a good idea?

**Comment by maximbaz** [comment]
I see.

Thumbnail mode behaves predictably and I like this very much (I can `tt`, move around, `Enter` in a pic, move with `n` a few times, then back to `tt` and the image I was viewing last is the one that is currently selected in thumbnail mode).

If library would also "follow" my selection like the thumbnail mode does, that would be very nice indeed, and consistent 👍 

Would it also be possible to synchronize thumbnail and library modes? Right now, if I move with `n` in the library mode, the selection in the thumbnail mode follows accordingly, but if my focus is in the thumbnail mode and I move around with `hjkl`, I'd like the library mode to also follow my cursor.

Basically I was to achieve a shared experience among the modes, if I am viewing a certain image, I want this image to be focused in all modes 🙂

**Comment by karlch** [comment]
This should be doable and makes sense to me  :)

**Comment by karlch** [comment]
Should be implemented in dc5c0a1cc0b5379179ede7efad4f50092bca7cdf and da0f023cea1818e42613daf3fc134a0716532a78. Keeping the issue open for now as I don't trust my late-night commits enough :smile:

**Comment by maximbaz** [comment]
Thank you @karlch, really appreciate your work!

It works well, I noticed a few inconsistencies that I'd like to bring up for the discussion, to double check if these are bugs or intended behavior 🙂 

1. When you navigate the library with `j/k`, the thumbnail mode follows the cursor, but the image mode does not. If you navigate the library with `n/p` shortcuts, both thumbnail and image mode follow the cursor.

     There probably is a reason why you have two sets of shortcuts (`j/k` and `n/p`) (what is it, performance?), if so should the behavior be consistent with thumbnail mode, i.e. `j/k` to not update thumbnail mode, but `n/p` to do so?

    Or is the reason for having `j/k` and `n/p` purely historical? In which case, maybe these keys should both do what `n/p` does today?

    Personally I will configure vimiv to do the latter (make `j/k` always synchronized), but I wonder if you want to do the same for everyone.

1. When you navigate the library with `j/k`, as said above the thumbnail mode follows the cursor. However if you change to a different folder, thumbnail mode will not contain new images, and further pressing `j/k` does nothing to thumbnail mode, it is stuck with the pictures from the previous dir. On contrast, if you navigate with `n/p` in the library mode, once you switch to a new directory and press `n/p`, the thumbnail mode gets updated with the new set of pictures and continues to follow the cursor.

    Just like in the previous point, I think the behavior of `n/p` is more predictable and personally I would make my `j/k` do the same, question is if you want to do this for everybody.

    An extra point is whether the moment of switching directories should immediately update the thumbnail mode, in other words, as soon as I entered the directory with `Enter`, can we make thumbnail mode already show images from this dir, before I even start moving around with `n/p`?

**Comment by karlch** [comment]
Thanks for all your detailed comments, feedback really helps!

For both 1 and 2 the reason is performance, loading images is expensive. I often find myself browsing the library just scrolling around until I find the image I am looking for which I then explicitly load by selecting it. Another thing that goes wrong horribly with `n/p` is just holding the keys in the library while scrolling up/down. I would therefore like to keep `j/k` as the default to keep scrolling and browsing the library performant.

The inconsistency you mention for 1, namely that thumbnail mode follows `j/k` but image mode does not, is something I was unsure about. If the images are already loaded into thumbnail mode, following the library is cheap. Loading new images on the other hand is always expensive. The idea was to synchronize as much as possible without large overhead with `j/k` and to fully synchronize with `n/p`. Not sure if `j/k` should not update thumbnail mode for consistency, even if this is cheap.

Concerning the final point when switching directories: the reason `n/p` works here is that behind the scenes the imagelist is updated when explicitly opening a new image. In principle scrolling right could also support `--open-selected` somehow, i.e. if we enter a new directory and an image is focused, open that image and load the new filelist. An alternative could be something along "if there are images in the new directory, load them". This would again hit performance though and may lead to funky corner-cases, e.g. when opening the home directory any images there would be opened, although this may just be an intermediate step. Happy to hear any opinions :)

**Comment by maximbaz** [comment]
> For both 1 and 2 the reason is performance, loading images is expensive... I would therefore like to keep j/k as the default to keep scrolling and browsing the library performant.

I see, makes sense to have it as default indeed 👍 

> The idea was to synchronize as much as possible without large overhead with j/k and to fully synchronize with n/p. Not sure if j/k should not update thumbnail mode for consistency, even if this is cheap.

On the one hand I like consistency, on the other hand I cannot really think of an example where I would _not_ want this synchronization. I'm okay with keeping it either way 🙂 

> In principle scrolling right could also support --open-selected somehow, i.e. if we enter a new directory and an image is focused, open that image and load the new filelist. An alternative could be something along "if there are images in the new directory, load them". This would again hit performance though and may lead to funky corner-cases, e.g. when opening the home directory any images there would be opened, although this may just be an intermediate step. Happy to hear any opinions :)

My main motivation here is that I find it slightly confusing that I have library and thumbnail modes open and they show contents of different directories.

For this reason I'd love to keep them in sync, so if I enter a directory, even if my cursor in the inner directory is not on an image, the thumbnail mode would be cleared and would begin to load images from this dir (can it be done asynchronously, not to block my input if I want to move around while thumbnails are still being loaded?).

**Comment by karlch** [comment]
> On the one hand I like consistency, on the other hand I cannot really think of an example where I would not want this synchronization.

I came to the same conclusion, so I guess we will just keep it like this.

> My main motivation here is that I find it slightly confusing that I have library and thumbnail modes open and they show contents of different directories.

This makes sense in general and could be possible. Maybe involving a setting to switch this off and on. I have some more questions on the expected behaviour though:
1. If there is no image in the directory, should thumbnail mode be cleared?
2. Should the first/selected image also be loaded behind the scenes for image mode? Probably has to be done (see later).

> can it be done asynchronously

Thumbnails themselves are already created and loaded in async, so this part should not be a problem. We also already know the list of images when entering the library, as the library shows them, so the part of figuring out the filelist is also done. What is not async currently, but should be possible, is the loading of an image for image mode. This goes hand in hand with 2. and would probably have to be done. Thumbnail mode very much relies on the fact, that image mode has the same filelist. Not sure what would break if this is changed. It would also be weird to toggle thumbnail mode, without selecting an image, i.e. `tt` or `gi`, and having a completely unrelated filelist in image mode as remnant.

**Comment by maximbaz** [comment]
> If there is no image in the directory, should thumbnail mode be cleared?

Yep, otherwise the confusion doesn't go away if thumbnail mode shows images from a different directory.

> Should the first/selected image also be loaded behind the scenes for image mode? Probably has to be done (see later).
> ...
> What is not async currently, but should be possible, is the loading of an image for image mode. This goes hand in hand with 2. and would probably have to be done. Thumbnail mode very much relies on the fact, that image mode has the same filelist. Not sure what would break if this is changed. It would also be weird to toggle thumbnail mode, without selecting an image, i.e. tt or gi, and having a completely unrelated filelist in image mode as remnant.

I agree, yes.

**Comment by karlch** [comment]
So we agree on the ideal implementation, nice. As this is quite far from the original issue, I have opened #132 to keep things organized. Since the current behaviour doesn't bother me much, I may not work on this anytime soon, sorry for that. Feel free to close this issue if the library part is solved :)

**Comment by maximbaz** [comment]
I think this part is solved, so closing, thanks again :)
